django/contrib/staticfiles/storage.py

```diff
@@ -224,39 +224,49 @@ def converter(matchobj):
             if re.match(r"^[a-z]+:", url):
                 return matched
 
-            # Ignore absolute URLs that don't point to a static file (dynamic
-            # CSS / JS?). Note that STATIC_URL cannot be empty.
-            if url.startswith("/") and not url.startswith(settings.STATIC_URL):
-                return matched
-
             # Strip off the fragment so a path-like fragment won't interfere.
             url_path, fragment = urldefrag(url)
 
             # Ignore URLs without a path
             if not url_path:
                 return matched
 
+            absolute_url_without_static_url = False
             if url_path.startswith("/"):
-                # Otherwise the condition above would have returned prematurely.
-                assert url_path.startswith(settings.STATIC_URL)
-                target_name = url_path.removeprefix(settings.STATIC_URL)
+                if url_path.startswith(settings.STATIC_URL):
+                    target_name = url_path[len(settings.STATIC_URL) :]
+                else:
+                    target_name = url_path[1:]
+                    absolute_url_without_static_url = True
             else:
                 # We're using the posixpath module to mix paths and URLs conveniently.
                 source_name = name if os.sep == "/" else name.replace(os.sep, "/")
                 target_name = posixpath.join(posixpath.dirname(source_name), url_path)
 
-            # Determine the hashed name of the target file with the storage backend.
-            hashed_url = self._url(
-                self._stored_name,
-                unquote(target_name),
-                force=True,
-                hashed_files=hashed_files,
-            )
+            try:
+                # Determine the hashed name of the target file with the storage backend.
+                hashed_url = self._url(
+                    self._stored_name,
+                    unquote(target_name),
+                    force=True,
+                    hashed_files=hashed_files,
+                )
+            except (ValueError, SuspiciousFileOperation):
```

Comment:
Do we have a test for `ValueError`? I don't see any failures in `staticfiles_tests` when I remove it.

Reply:
This block of code has some redundancy. ValueError is only raised when the URL does not start with a '/' and thus is always reraised, and SuspiciousFileOperation is always raised when it does start with a '/'. Removing all redundant code would result in:

```
except SuspiciousFileOperation:
    # Ignore absolute URLs that don't point to a static file (dynamic
    # CSS / JS?). Note that STATIC_URL cannot be empty.
    if not url.startswith(settings.STATIC_URL):
        return matched
    else:
        raise
```

And it would be all fine. 

However, I prefer to have some redundancy as I am just not 100% relying on the code this comes from. The exception originates in django/utils/_os.py and although the docstring claims it should raise a ValueError the code actually raises a SuspiciousFileOperation.

Reply:
I think it should be fine to remove `ValueError`. If you look a very early commit eg f6e86c4286 then you can see the function did actually raise a `ValueError`. When it was changed to `SuspiciousFileOperation` the docstring wasn't updated 🤷‍♂️

Reply:
ah confirmed: b8ba73cd0cb6a3dbdaeb3df65936970956829de3

/me looks at aaugistin 😁

Reply:
IMO, `ValueError` is still needed, e.g.
```diff
diff --git a/tests/staticfiles_tests/project/documents/cached/source_map.js b/tests/staticfiles_tests/project/documents/cached/source_map.js
index 9d417868a0..10391168df 100644
--- a/tests/staticfiles_tests/project/documents/cached/source_map.js
+++ b/tests/staticfiles_tests/project/documents/cached/source_map.js
@@ -1,2 +1,3 @@
 //# sourceMappingURL=source_map.js.map
+//# sourceMappingURL=/source_map.js.map
 let a_variable = 1;
```
crashes with:
```
ValueError: The file 'source_map.js.map' could not be found with <django.contrib.staticfiles.storage.ManifestStaticFilesStorage object at 0x7fab98dd2fe0>.
```

Reply:
Good one. Yes, there seemed to be missing a test that tests the behavior of "Ignore absolute URLs that don't point to a static file". Now pushed.